pkg/lifecycle/service.go

```diff
@@ -176,11 +242,13 @@ func (s *Service) stopForceful(ctx context.Context, pl *Instance) error {
 // StopAll will ask all the pipelines to stop gracefully
 // (i.e. that existing messages get processed but not new messages get produced).
 func (s *Service) StopAll(ctx context.Context, reason error) {
-	for _, pl := range s.instances {
-		if pl.GetStatus() != StatusRunning && pl.GetStatus() != StatusRecovering {
+	instances := s.pipelines.List(ctx)
+	for _, pl := range instances {
+		rp, ok := s.runningPipelines[pl.ID]
```

Comment:
Again here, we need to hold the lock if we are accessing `s.runningPipelines`. Also, I'm not sure we even need to list the pipelines, we should be able to just range over the ones in `s.runningPipelines`.

Reply:
Good point. This will be a way to reutilize the runningPipelines more effectively.